src/CommandLineIface.cpp

```diff
@@ -0,0 +1,90 @@
+                                                                                   settings_(this), translator_(new MarianInterface(this)),
+                                                                                   qcin_(stdin), qcout_(stdout), qcerr_(stderr) {
+    connect(translator_, &MarianInterface::error, this, &CommandLineIface::outputError);
+    connect(translator_, &MarianInterface::translationReady, this, &CommandLineIface::outputTranslation);
+}
+
+int CommandLineIface::run() {
+    if (parser_.isSet("l")) {
+        printLocalModels();
+        return 0;
+    } else if (parser_.isSet("m")) {
+        QString model_shortname = parser_.value("model");
+
+        // Try to find our model in the list of models
+        QString modelpath("");
+        for (auto&& model : models_.getInstalledModels()) {
+            if (model.shortName == model_shortname) {
+                modelpath = model.path;
+            }
+        }
+        if (modelpath == "") {
+            qcerr_ << "We could not find a model identified as: " << model_shortname << " . Use translateLocally -l to list available models or use the GUI to download some from the internet.\n";
+            return 1;
+        }
+        // Init the translation model
+        translator_->setModel(modelpath, settings_.marianSettings());
+        doTranslation();
+        return 0;
+    } else {
+        qcerr_ << "We are in command line mode, but there's nothing for us to do. Some control flow mistake maybe?\n";
+        return 2;
+    }
+}
+
+void CommandLineIface::printLocalModels() {
+    QList<Model> localmodels = models_.getInstalledModels();
+    for (auto&& model : localmodels) {
+        qcout_ << model.src << "-" << model.trg << " type: " << model.type << " version: " << model.localversion << "; To invoke do -m " << model.shortName << '\n';
+    }
+}
+
+/* This function is pseudo blocking, via an event loop.*/
+void CommandLineIface::doTranslation() {
+    // Find whether input is stdin or a file
+    QString input;
+    if (parser_.isSet("i")) {
+        QFile inputFile(parser_.value("i"));
+        if (inputFile.open(QIODevice::ReadOnly)) {
+           QTextStream in(&inputFile);
+           input = in.readAll(); // @TODO this should be more complicated batching thingie, but alas. Should work fine with most small inputs.
+                                 // Alternative is to read it line by line but then it would be horribly inefficient as we would lose all batching.
+           inputFile.close();
+        } else {
+            outputError(QString("Couldn't open input file: " + parser_.value("i")));
+        }
+    } else {
+        input = qcin_.readAll(); // @TODO again this should do some batching and allow for full cmd usage but I *really* don't want to re-implement marian...
+                                 // Maybe expose the raw marian interface with their batcher, or the service which has some sort of CLI usage.
```

Comment:
I'm tempted to actively push that kind of use case to either Marian itself, or bergamot-translator's (upcoming?) Python module.

Reply:
Do you mean the whole CMD interface or proper batching? I should be able to write a short batcher with not too much difficulty, but I don't want to recreate functionality

Reply:
(In case it was a bit lost in Gather.town today) My idea is to have translateLocally be able to do command line translation, but not be best-in-class in it. If they want speed and efficiency, they should go to Marian. If they want to embed it, either marian-server or the Python interface of bergamot. If you they want to translate a file, then, okay. But I am assuming those files are a couple of MBs at most. Not GBs.

However, if a simple loop (i.e. `while file not empty; read 1000 lines, translate (with batching inside those 1000 lines handled by bergamot transparently) print translations`) can already prevent people from swamping their memory and get what they want, that would be a worthy addition. *But the goal should not be translate very large files as fast as possible, just translate large files without running out of memory*.

Additionally, ideally the translatelocally.com should have a little section on alternatives (such as marian-decoder and bergamot's Python API) for advanced use cases, hidden away for those who search for it.